File: longest-common-subsequence/lcs.php
<?php
function lcs($X, $Y) {
    $m = strlen($X);
    $n = strlen($Y);
    $L = array_fill(0, $m + 1, array_fill(0, $n + 1, 0));
    
    for ($i = 0; $i <= $m; $i++) {
        for ($j = 0; $j <= $n; $j++) {
            if ($i == 0 || $j == 0) {
                $L[$i][$j] = 0;
            } elseif ($X[$i-1] == $Y[$j-1]) {
                $L[$i][$j] = $L[$i-1][$j-1] + 1;
            } else {
                $L[$i][$j] = max($L[$i-1][$j], $L[$i][$j-1]);
            }
        }
    }
    
    return $L[$m][$n];
}

echo lcs("ABCDGH", "AEDFHR");
?>

File: longest-common-subsequence/LCS.kt
fun lcs(X: String, Y: String): Int {
    val m = X.length
    val n = Y.length
    val L = Array(m + 1) { IntArray(n + 1) }
    
    for (i in 0..m) {
        for (j in 0..n) {
            when {
                i == 0 || j == 0 -> L[i][j] = 0
                X[i-1] == Y[j-1] -> L[i][j] = L[i-1][j-1] + 1
                else -> L[i][j] = maxOf(L[i-1][j], L[i][j-1])
            }
        }
    }
    
    return L[m][n]
}

fun main() {
    println(lcs("ABCDGH", "AEDFHR"))
}
